### PR TITLE
[SYCL-MLIR]: Uniformity analysis (part1)

### PR DESCRIPTION
**Pull request by etiotto**
Introduce an analysis to determine whether a value is `uniform` or not. A `uniform` value is a value that yields the same result for all thread.

### CODE REVIEW

File: polygeist/lib/Dialect/Polygeist/Analysis/CMakeLists.txt [hunk 1]
```diff
@@ -1,5 +1,6 @@
 add_mlir_dialect_library(MLIRPolygeistAnalysis
   MemoryAccessAnalysis.cpp
+  UniformityAnalysis.cpp
   ReachingDefinitionAnalysis.cpp
```
Comment (whitneywhtsang) [inline polygeist/lib/Dialect/Polygeist/Analysis/CMakeLists.txt:4]:
```suggestion
  ReachingDefinitionAnalysis.cpp
  UniformityAnalysis.cpp
```

File: polygeist/lib/Dialect/Polygeist/Analysis/UniformityAnalysis.cpp [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+} // namespace polygeist
+} // namespace mlir
+
+void UniformityLattice::onUpdate(DataFlowSolver *solver) const {
+  Lattice::onUpdate(solver);
+
+  auto value = point.get<Value>();
+  Operation *op = value.getDefiningOp();
+
+  if (op) {
```
Comment (whitneywhtsang) [inline polygeist/lib/Dialect/Polygeist/Analysis/UniformityAnalysis.cpp:55]:
```suggestion
  if (Operation *op = value.getDefiningOp();) {
```

File: polygeist/test/lib/Dialect/Polygeist/TestUniformityAnalysis.cpp [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+  void runOnOperation() override {
+    DataFlowSolver solver;
+    solver.load<DeadCodeAnalysis>();
+    solver.load<UniformityAnalysis>();
+
+    Operation *op = getOperation();
+    if (failed(solver.initializeAndRun(op)))
+      return signalPassFailure();
+
+    auto printUniformity = [](const Uniformity &info, raw_ostream &os) {
```
Comment (whitneywhtsang) [inline polygeist/test/lib/Dialect/Polygeist/TestUniformityAnalysis.cpp:41]: Given that you defined `raw_ostream &operator<<(raw_ostream &os, const Uniformity &uniformity)`, can we use `os << info;` ?

File: polygeist/test/lib/Dialect/Polygeist/TestUniformityAnalysis.cpp [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+      const UniformityLattice *uniformity =
+          solver.lookupState<UniformityLattice>(op->getResult(0));
+      assert(uniformity && "expected uniformity information");
+      assert(!uniformity->getValue().isUninitialized() &&
+             "lattice element should be initialized");
+
+      llvm::errs() << tag.getValue() << ", " << *uniformity << "\n";
+
+      //      printUniformity(uniformity->getValue(), llvm::errs());
```
Comment (whitneywhtsang) [inline polygeist/test/lib/Dialect/Polygeist/TestUniformityAnalysis.cpp:68]: remove.

File: polygeist/test/polygeist-opt/uniformity.mlir [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+!sycl_id_2 = !sycl.id<[2], (!sycl_array_2)>
+!sycl_range_2 = !sycl.range<[2], (!sycl_array_2)>
+!sycl_accessor_impl_device_2 = !sycl.accessor_impl_device<[2], (!sycl_id_2, !sycl_range_2, !sycl_range_2)>
+!sycl_group_2 = !sycl.group<[2], (!sycl_range_2, !sycl_range_2, !sycl_range_2, !sycl_id_2)>
+!sycl_item_base_2 = !sycl.item_base<[2, true], (!sycl_range_2, !sycl_id_2, !sycl_id_2)>
+!sycl_accessor_2_f32_r_gb = !sycl.accessor<[2, f32, read, global_buffer], (!sycl_accessor_impl_device_2, !llvm.struct<(memref<?xf32, 2>)>)>
+!sycl_item_2 = !sycl.item<[2, true], (!sycl_item_base_2)>
+!sycl_nd_item_2 = !sycl.nd_item<[2], (!sycl_item_2, !sycl_item_2, !sycl_group_2)>
+
+// CHECK: test1_add1, uniformity: unknown
```
Comment (whitneywhtsang) [inline polygeist/test/polygeist-opt/uniformity.mlir:13]: I foresee this function is going to grow, how about inline the CHECK in function body, so they are easier to reference?

File: polygeist/test/polygeist-opt/uniformity.mlir [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+// CHECK: test1_add4, uniformity: uniform
+func.func @test1(%arg0 : i1, %arg1: memref<?x!sycl_nd_item_2>)  {
+  %c0_i32 = arith.constant 0 : i32  
+  %true = arith.constant 1 : i1
+  %c2 = arith.constant 2 : i64
+  %c3 = arith.constant 2 : i64
+  %tx = sycl.nd_item.get_global_id(%arg1, %c0_i32) : (memref<?x!sycl_nd_item_2>, i32) -> i64
+
+  // %arg0 uniformity is unknown -> %v0 uniformity also unknown
+
```
Comment (whitneywhtsang) [inline polygeist/test/polygeist-opt/uniformity.mlir:25]: nit: remove empty line.

File: polygeist/test/polygeist-opt/uniformity.mlir [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+  // branch condition is non-uniform -> %v2 is non-uniform
+  %cond2 = arith.cmpi slt, %tx, %c1 : i64
+  %v2 = scf.if %cond2 -> i64 {
+    scf.yield %c2 : i64
+  } else {
+    scf.yield %c3 : i64
+  }
+  %add2 = arith.addi %v2, %c1 {tag = "test1_add2"} : i64  
+
+  // branch condition is uniform, but yielded value is non-uniform -> %v0 is non-uniform
```
Comment (whitneywhtsang) [inline polygeist/test/polygeist-opt/uniformity.mlir:43]:
```suggestion
  // branch condition is uniform, but yielded value is non-uniform -> %v3 is non-uniform
```

File: polygeist/test/polygeist-opt/uniformity.mlir [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+  %v3 = scf.if %true -> i64 {
+    scf.yield %tx : i64
+  } else {
+    scf.yield %c3 : i64
+  }  
+  %add3 = arith.addi %v3, %c1 {tag = "test1_add3"} : i64
+
+  // branch condition is uniform, and yielded values is uniform -> %v0 is uniform
```
Comment (whitneywhtsang) [inline polygeist/test/polygeist-opt/uniformity.mlir:51]:
```suggestion
  // branch condition is uniform, and yielded values is uniform -> %v4 is uniform
```

File: polygeist/lib/Dialect/Polygeist/Analysis/UniformityAnalysis.cpp [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+
+  // If the lattice on any operand is uninitialized, bail out.
+  if (llvm::any_of(operands, [](const UniformityLattice *lattice) {
+        return lattice->getValue().isUninitialized();
+      })) {
+    return;
+  }
+
+  // These operations always yield non-uniform result(s).
+  if (isa<sycl::SYCLItemGetIDOp, sycl::SYCLNDItemGetGlobalIDOp>(op)) {
```
Comment (sommerlukas) [inline polygeist/lib/Dialect/Polygeist/Analysis/UniformityAnalysis.cpp:92]: Does this cover all relevant cases, including kernels only taking an `id` as argument, i.e.
```
cgh.parallel_for(range<1>{...}, [=](id<1> i){...});
```

File: polygeist/lib/Dialect/Polygeist/Analysis/UniformityAnalysis.cpp [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+  assert(llvm::all_of(operands,
+                      [&](const UniformityLattice *lattice) {
+                        return lattice->getValue().getKind() ==
+                               Uniformity::Kind::Uniform;
+                      }) &&
+         "Expecting all operands to be uniform");
+
+  // A memory side effects free operation that has uniform operands yields
+  // uniform result(s).
+  if (isMemoryEffectFree(op)) {
```
Comment (victor-eds) [inline polygeist/lib/Dialect/Polygeist/Analysis/UniformityAnalysis.cpp:125]: This is incorrect, as there are memory effect free operations with non-uniform results, e.g., [`gpu.thread_id`](https://mlir.llvm.org/docs/Dialects/GPU/#gputhread_id-gputhreadidop).

File: polygeist/lib/Dialect/Polygeist/Analysis/UniformityAnalysis.cpp (hunk 9 above)
Comment (victor-eds) [reply polygeist/lib/Dialect/Polygeist/Analysis/UniformityAnalysis.cpp:92]: I think we shouldn't be handling any operation differently here. I propose creating a `UniformityOpInterface` for operations to express whether they define uniform values.

Comment (etiotto) [reply polygeist/lib/Dialect/Polygeist/Analysis/UniformityAnalysis.cpp:92]: > Maybe the correct way to approach this is to assume an operation does not define non-uniform values unless it implements some interface.

Yes we need to assume that operations in other dialects (e.g. an arith constant definition) yield uniform values

> We could then attach this interface to all operations that we know will produce a non-uniform value in the SYCL dialect and other relevant dialects and then basically do taint-tracking, where the "taint" is non-uniform values

Yes, essentially certain operations *always* define non-uniform values even if the input operands (if any) are uniform.  

I am willing to add an interface to our SYCL dialect operations, either in this PR or a subsequent one.

File: polygeist/lib/Dialect/Polygeist/Analysis/UniformityAnalysis.cpp (hunk 10 above)
Comment (etiotto) [reply polygeist/lib/Dialect/Polygeist/Analysis/UniformityAnalysis.cpp:125]: Right, the idea is to catch those operations at line 92 (like we do for the SYCL operations that yield a thread id). I didn't address other possible sources of non-uniformity in this PR (is part 1:-)

File: polygeist/lib/Dialect/Polygeist/Analysis/CMakeLists.txt (hunk 1 above)
Comment (etiotto) [reply polygeist/lib/Dialect/Polygeist/Analysis/CMakeLists.txt:4]: ty - done

File: polygeist/lib/Dialect/Polygeist/Analysis/UniformityAnalysis.cpp (hunk 2 above)
Comment (etiotto) [reply polygeist/lib/Dialect/Polygeist/Analysis/UniformityAnalysis.cpp:55]: done

File: polygeist/test/lib/Dialect/Polygeist/TestUniformityAnalysis.cpp (hunk 3 above)
Comment (etiotto) [reply polygeist/test/lib/Dialect/Polygeist/TestUniformityAnalysis.cpp:41]: Yes, I forgot to remove this one. We are using the `operator<<` you mention (line 66). Cleaning this up now.

File: polygeist/test/lib/Dialect/Polygeist/TestUniformityAnalysis.cpp (hunk 4 above)
Comment (etiotto) [reply polygeist/test/lib/Dialect/Polygeist/TestUniformityAnalysis.cpp:68]: done

File: polygeist/test/polygeist-opt/uniformity.mlir (hunk 6 above)
Comment (etiotto) [reply polygeist/test/polygeist-opt/uniformity.mlir:25]: done

File: polygeist/test/polygeist-opt/uniformity.mlir (hunk 7 above)
Comment (etiotto) [reply polygeist/test/polygeist-opt/uniformity.mlir:43]: modified the test a bit and addressed this comment.

File: polygeist/test/polygeist-opt/uniformity.mlir (hunk 8 above)
Comment (etiotto) [reply polygeist/test/polygeist-opt/uniformity.mlir:51]: modified the test a bit and addressed this comment.

File: polygeist/test/polygeist-opt/uniformity.mlir (hunk 5 above)
Comment (etiotto) [reply polygeist/test/polygeist-opt/uniformity.mlir:13]: Agree.

File: polygeist/lib/Dialect/Polygeist/Analysis/UniformityAnalysis.cpp (hunk 9 above)
Comment (sommerlukas) [reply polygeist/lib/Dialect/Polygeist/Analysis/UniformityAnalysis.cpp:91]: > I am willing to add an interface to our SYCL dialect operations, either in this PR or a subsequent one.

Sounds like a good approach, I leave it up to you if you want to add the interface in this PR or in the next part of the analysis.

**Review by sommerlukas — APPROVED** [review]

**Review by whitneywhtsang — APPROVED** [review]
LGTM as a part 1 PR.

**Pull request merged**